File: src/prompts/profiler_topics.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Union



@dataclass
class ProfileTopic:
    topic: str
    sub_topics: List[Union[str, Dict[str, str]]] = field(default_factory=list)
    description: str = ""


PROFILE_TOPICS: List[ProfileTopic] = [
    ProfileTopic(
        "basic_info",
        sub_topics=[
            "Name",
            {"name": "Age", "description": "integer"},
            "Gender",
            "birth_date",
            "nationality",
            "ethnicity",
            "language_spoken",
        ],
    ),
    ProfileTopic(
        "contact_info",
        sub_topics=["email", "phone", "city", "country"],
    ),
    ProfileTopic(
        "education",
        sub_topics=["school", "degree", "major"],
    ),
    ProfileTopic(
        "demographics",
        sub_topics=["marital_status", "spouse_name", "number_of_children", "household_income", "relationship"],
    ),
    ProfileTopic(
        "work",
        sub_topics=["company", "title", "working_industry", "previous_projects", "work_skills"],
    ),
    ProfileTopic(
        "interest",
        sub_topics=["books", "movies", "music", "foods", "sports", "hobbies", "art", "travel", "games"],
    ),
    ProfileTopic(
        "life_event",
        sub_topics=["marriage", "relocation", "retirement", "health", "achievement"],
    ),
]


def format_topics_for_prompt(topics: List[ProfileTopic] | None = None) -> str:
    topics = topics or PROFILE_TOPICS
    lines: list[str] = []
    for t in topics:
        desc = f" ({t.description})" if t.description else ""
        lines.append(f"- {t.topic}{desc}")
        for st in t.sub_topics:
            if isinstance(st, dict):
                st_desc = f"({st.get('description', '')})" if st.get("description") else ""
                lines.append(f"  - {st['name']}{st_desc}")
            else:
                lines.append(f"  - {st}")
    lines.append("...")
    return "\n".join(lines)

File: xmem-go/internal/prompts/profiler_topics.go
package prompts

import (
	"fmt"
	"strings"
)

type SubTopic struct {
	Name        string
	Description string
}

type ProfileTopic struct {
	Topic       string
	SubTopics   []SubTopic
	Description string
}

var ProfileTopics = []ProfileTopic{
	{
		Topic: "basic_info",
		SubTopics: []SubTopic{
			{Name: "Name"},
			{Name: "Age", Description: "integer"},
			{Name: "Gender"},
			{Name: "birth_date"},
			{Name: "nationality"},
			{Name: "ethnicity"},
			{Name: "language_spoken"},
		},
	},
	{
		Topic: "contact_info",
		SubTopics: []SubTopic{
			{Name: "email"},
			{Name: "phone"},
			{Name: "city"},
			{Name: "country"},
		},
	},
	{
		Topic: "education",
		SubTopics: []SubTopic{
			{Name: "school"},
			{Name: "degree"},
			{Name: "major"},
		},
	},
	{
		Topic: "demographics",
		SubTopics: []SubTopic{
			{Name: "marital_status"},
			{Name: "spouse_name"},
			{Name: "number_of_children"},
			{Name: "household_income"},
			{Name: "relationship"},
		},
	},
	{
		Topic: "work",
		SubTopics: []SubTopic{
			{Name: "company"},
			{Name: "title"},
			{Name: "working_industry"},
			{Name: "previous_projects"},
			{Name: "work_skills"},
		},
	},
	{
		Topic: "interest",
		SubTopics: []SubTopic{
			{Name: "books"},
			{Name: "movies"},
			{Name: "music"},
			{Name: "foods"},
			{Name: "sports"},
			{Name: "hobbies"},
			{Name: "art"},
			{Name: "travel"},
			{Name: "games"},
		},
	},
	{
		Topic: "life_event",
		SubTopics: []SubTopic{
			{Name: "marriage"},
			{Name: "relocation"},
			{Name: "retirement"},
			{Name: "health"},
			{Name: "achievement"},
		},
	},
}

func FormatTopicsForPrompt(topics []ProfileTopic) string {
	if topics == nil {
		topics = ProfileTopics
	}
	var lines []string
	for _, t := range topics {
		desc := ""
		if t.Description != "" {
			desc = fmt.Sprintf(" (%s)", t.Description)
		}
		lines = append(lines, fmt.Sprintf("- %s%s", t.Topic, desc))
		for _, st := range t.SubTopics {
			stDesc := ""
			if st.Description != "" {
				stDesc = fmt.Sprintf("(%s)", st.Description)
			}
			lines = append(lines, fmt.Sprintf("  - %s%s", st.Name, stDesc))
		}
	}
	lines = append(lines, "...")
	return strings.Join(lines, "\n")
}
